File: dpcsh/benchmark/restful_dpc.go
// Copyright 2020 Fungible Inc

package main

import (
	"bytes"
	"encoding/json"
	"flag"
	"log"
	"math"
	"net/http"
)

var (
	serviceAddr               = flag.String("service_addr", ":5000", "http service address")
	dpcProto                  = flag.String("dpc_protocol", "tcp", "DPC proxy protocol")
	dpcAddr                   = flag.String("dpc_addr", "localhost:40221", "DPC proxy address")
	certFile                  = flag.String("cert_file", "cert.pem", "TLS certificate")
	keyFile                   = flag.String("key_file", "key.pem", "TLS key")
	responseSize              = flag.Int("response_size", 207, "Response size in bytes")
	tlsEnable                 = flag.Bool("tls_enable", false, "Enable TLS")
	client         *DpcClient = nil
	requestsServed            = 0
)

type storageCommandParams struct {
	BlockSize int    `json:"block_size"`
	Capacity  int    `json:"capacity"`
	GroupID   int    `json:"group_id"`
	Name      string `json:"name"`
	Type      string `json:"type"`
	UUID      string `json:"uuid"`
}

type storageCommand struct {
	Class  string               `json:"class"`
	OpCode string               `json:"opcode"`
	Params storageCommandParams `json:"params"`
}

func serveHome(w http.ResponseWriter, r *http.Request) {
	if r.Method != "GET" {
		http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
		return
	}
	// useful code here
	params := storageCommand{"volume", "VOL_ADMIN_OPCODE_CREATE", storageCommandParams{4096, 8589934592, 1, "test1", "VOL_TYPE_BLK_LOCAL_THIN", "a8c6820a8b0b3ef1"}}
	response, err := client.Execute("echo", []interface{}{params})

	if err != nil {
		http.Error(w, "Internal server error", http.StatusInternalServerError)
		return
	}
	jsonStr, err := json.Marshal(response)
	if err != nil {
		http.Error(w, "Internal server error", http.StatusInternalServerError)
		return
	}
	w.Write(bytes.Repeat(jsonStr, int(math.Ceil(float64(*responseSize)/float64(len(jsonStr)))))[:*responseSize])
	requestsServed++
	if requestsServed%1000 == 0 {
		log.Println(r.URL)
		log.Printf("Served %d requests", requestsServed)
	}
}

func main() {
	flag.Parse()
	log.SetFlags(0)
	var err error
	client, err = NewClient(*dpcProto, *dpcAddr)
	if err != nil {
		log.Fatal(err)
		return
	}
	http.HandleFunc("/", serveHome)
	if *tlsEnable {
		log.Fatal(http.ListenAndServeTLS(*serviceAddr, *certFile, *keyFile, nil))
	} else {
		log.Fatal(http.ListenAndServe(*serviceAddr, nil))
	}
}

File: dpcsh/benchmark/restful_dpc.py
#!/usr/bin/env python3
"""Usage: restful_dpc.py [-h] [--unix_socket=<file>] [--response-size=<bytes>]
          [--tcp_port=<port>] [--https]

Provides restful DPC-echo service via Flask.
"""

from docopt import docopt
import flask
from dpc_client import DpcClient
import json
import math
import sys
from gevent.pywsgi import WSGIServer

app = flask.Flask(__name__)
client = None
response_size = None

@app.route('/', methods=["GET", "POST"])
def echo():
  global client
#  if not flask.request.json:
#    print("Bad json format")
#    flask.abort(400)

  result = client.execute('echo', [
        {
            "class": "volume", 
            "opcode": "VOL_ADMIN_OPCODE_CREATE", 
            "params": {
                "block_size": 4096, 
                "capacity": 8589934592, 
                "group_id": 1, 
                "name": "test1", 
                "type": "VOL_TYPE_BLK_LOCAL_THIN", 
                "uuid": "a8c6820a8b0b3ef1"
            }
        }
    ], None)
  s = json.dumps(result)
  return s * (int(response_size / len(s))) + s[:(response_size % len(s))]

def main():
  global client, response_size
  arguments = docopt(__doc__)
  if arguments['--tcp_port'] is None and arguments['--unix_socket'] is None:
    print('tcp_port or unix_socket is required')
    print(__doc__)
    sys.exit(1)

  response_size = 207 if arguments['--response-size'] is None else int(arguments['--response-size'])

  if arguments['--tcp_port'] is None:
    print('Running on Unix Socket', arguments['--unix_socket'])
    client = DpcClient(server_address=arguments['--unix_socket'], unix_sock=True)
  else:
    print('Running on TCP localhost:', arguments['--tcp_port'])
    client = DpcClient(server_address=('localhost', int(arguments['--tcp_port'])))

  print("Connected")

  if arguments['--https']:
    server = WSGIServer(
            ('0.0.0.0', 5000),
            app,
            log='default',
            keyfile='key.pem',
            certfile='cert.pem'
        )
    server.serve_forever()
  else:
    server = WSGIServer(
            ('0.0.0.0', 5000),
            app,
            log='default'
        )
    server.serve_forever()

if __name__ == "__main__":
    main()
